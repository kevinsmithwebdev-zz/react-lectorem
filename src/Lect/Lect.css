.Lect {
  font-size: 24px;
  font-weight: bold;
  line-height: 150%;
  white-space: pre-wrap;
}

.unHighlighted {
  color: black;
}

.highlighted {
  color: red;
}
