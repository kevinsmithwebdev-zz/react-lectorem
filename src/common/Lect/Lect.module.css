.Lect {
  font-size: 24px;
  font-weight: bold;
  white-space: pre-wrap;
}

.unHighlighted {
  color: black;
}

.highlighted {
  color: red;
}

.clickable {
  cursor: pointer;
}
