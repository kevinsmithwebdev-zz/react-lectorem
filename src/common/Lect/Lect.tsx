import React from 'react';
import classes from './Lect.module.css';
import { getFrontSpacer } from '../../utils/punctuation';
import { LectInterface } from '../../interfaces/index';

interface LectComponentInterface {
  lect: LectInterface,
  shouldHighlight: boolean,
  showModal: Function,
  isLast: boolean,
  isFirst: boolean,
  sentenceTranslation: string | undefined,
  originalSentence: string,
}

export const getHasModalInformation = (lect: LectInterface): boolean => !!lect.explanation || !!lect.translation;

export const getClassName = (shouldHighlight: boolean, hasModalInformation: boolean): string =>
  [
    classes.Lect,
    shouldHighlight ? classes.highlighted : classes.unHighlighted,
    hasModalInformation ? classes.clickable : '',
  ].join(' ');

const Lect: React.FC<LectComponentInterface> = ({
  lect, shouldHighlight, showModal, isLast, isFirst, sentenceTranslation, originalSentence,
}) => {
  const hasModalInformation = getHasModalInformation(lect);
  const className = getClassName(shouldHighlight, hasModalInformation);
  const frontSpacer = getFrontSpacer(lect.text[0], isFirst);

  const handleClick = hasModalInformation
    ? /* istanbul ignore next */ () => showModal({ ...lect, sentenceTranslation, originalSentence })
    : /* istanbul ignore next */ () => {};
  return (
    <span className={ className } id='howard' onClick={ handleClick } aria-hidden='true'>
      {`${frontSpacer}${lect.text}${isLast ? '  ' : ''}`}
    </span>
  );
};

export default Lect;
