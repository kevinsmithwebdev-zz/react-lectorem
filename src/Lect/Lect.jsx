import React from 'react';
import './Lect.css';

const endingPunctuation = ['.', '!', '?'];

const Lect = ({ lect: { text }, shouldHighlight, readTime }) => {
  const className = [
    'Lect',
    shouldHighlight ? 'highlighted' : 'unHighlighted',
  ].join(' ');
  const lastChar = text.slice(-1);
  const isEndingInPunctuation = endingPunctuation.includes(lastChar);
  const spacer = isEndingInPunctuation ? '  ' : ' ';
  return (
    <span className={ className } >
      { `${text}${spacer}` }
    </span>
  );
};

export default Lect;
