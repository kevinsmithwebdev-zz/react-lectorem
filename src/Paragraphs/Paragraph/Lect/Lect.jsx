import React from 'react';
import './Lect.css';

export const ENDING_PUNCTUATION = ['.', '!', '?'];
export const getSpacer = text => (ENDING_PUNCTUATION.includes(text?.slice(-1)) ? '  ' : ' ');
export const getHasModalInformation = lect => !!lect.explanation || !!lect.translation;
export const getClassName = (shouldHighlight, hasModalInformation) => [
  'Lect',
  shouldHighlight ? 'highlighted' : 'unHighlighted',
  hasModalInformation ? 'clickable' : '',
].join(' ');

const Lect = ({ lect, shouldHighlight, showModal }) => {
  if (!lect) {
    return null;
  }

  const hasModalInformation = getHasModalInformation(lect);
  const className = getClassName(shouldHighlight, hasModalInformation);
  const spacer = getSpacer(lect.text);

  const handleClick = hasModalInformation && /* istanbul ignore next */ (() => showModal(lect));

  return (
    <span
      className={ className }
      onClick={ handleClick }
      aria-hidden='true'
    >
      { `${lect.text}${spacer}` }
    </span>
  );
};

export default Lect;
