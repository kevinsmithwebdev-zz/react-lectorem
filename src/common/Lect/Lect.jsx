import React from 'react';
import './Lect.css';
import { getFrontSpacer } from '../../utils/punctuation';

export const getHasModalInformation = lect => !!lect.explanation || !!lect.translation;

export const getClassName = (shouldHighlight, hasModalInformation) =>
  [
    'Lect',
    shouldHighlight ? 'highlighted' : 'unHighlighted',
    hasModalInformation ? 'clickable' : '',
  ].join(' ');

const Lect = ({
  lect, shouldHighlight, showModal, isLast, isFirst, sentenceTranslation, originalSentence,
}) => {
  if (!lect) {
    return null;
  }
  const hasModalInformation = getHasModalInformation(lect);
  const className = getClassName(shouldHighlight, hasModalInformation);
  const frontSpacer = getFrontSpacer(lect.text[0], isFirst);

  const handleClick = hasModalInformation
    ? () => showModal({ ...lect, sentenceTranslation, originalSentence })
    : () => {};

  return (
    <span className={ className } onClick={ handleClick } aria-hidden='true'>
      {`${frontSpacer}${lect.text}${isLast ? '  ' : ''}`}
    </span>
  );
};

export default Lect;
