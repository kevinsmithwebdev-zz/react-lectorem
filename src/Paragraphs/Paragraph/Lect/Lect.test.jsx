import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Lect, { getSpacer, getHasModalInformation, getClassName, ENDING_PUNCTUATION } from './Lect';
import storyData from '../../../../fixture/storyData.json';

const mockLect = storyData.paragraphs[0].lects[0];

describe('Lect', () => {
  describe('the main component render', () => {
    const props = {
      lect: mockLect,
      showModal: 'showModal123',
    };

    const renderer = new ShallowRenderer();
    describe('empty', () => {
      it('should render without crashing', () => {
        renderer.render(<Lect />);
        expect(renderer.getRenderOutput()).toMatchSnapshot();
      });
    });

    describe('with data, should highlight', () => {
      it('should render and match snapshot', () => {
        renderer.render(<Lect { ...props } shouldHighlight />);
        expect(renderer.getRenderOutput()).toMatchSnapshot();
      });
    });

    describe('with data, should not highlight', () => {
      it('should render and match snapshot', () => {
        renderer.render(<Lect { ...props } shouldHighlight={ false } />);
        expect(renderer.getRenderOutput()).toMatchSnapshot();
      });
    });
  });

  describe('getSpacer', () => {
    describe('when ending in terminal punctuation', () => {
      ENDING_PUNCTUATION.forEach(pun => {
        describe(`when ending punctuation is "${pun}"`, () => {
          const text = `howdy${pun}`;
          it('should return two spaces', () => {
            expect(getSpacer(text)).toBe('  ');
          });
        });
      });
    });

    describe('when not ending in punctuation', () => {
      const text = 'howdy';
      it('should return one space', () => {
        expect(getSpacer(text)).toBe(' ');
      });
    });
  });

  describe('getHasModalInformation', () => {
    const testData = [
      { lect: { explanation: undefined, translation: undefined }, result: false },
      { lect: { explanation: undefined, translation: 'something' }, result: true },
      { lect: { explanation: 'something', translation: undefined }, result: true },
      { lect: { explanation: 'something', translation: 'something' }, result: true },
    ];
    const getLectPropText = (lect, prop) => lect[prop] ? `has ${prop}` : `does not have ${prop}`;

    testData.forEach(({ lect, result}) => {
      describe(`when ${getLectPropText(lect, 'explanation')} and ${getLectPropText(lect, 'translation')}`, () => {
        it(`should return ${result}`, () => {
          expect(getHasModalInformation(lect)).toBe(result);
        });
      });
    });
  });

  describe('getClassName', () => {
    const testData = [
      { flags: [false, false], result: 'Lect unHighlighted ' },
      { flags: [false, true], result: 'Lect unHighlighted clickable' },
      { flags: [true, false], result: 'Lect highlighted ' },
      { flags: [true, true], result: 'Lect highlighted clickable' },
    ];

    testData.forEach(({ flags: [shouldHighlight, hasModalInformation], result}) => {
      describe(`when shouldHighlight is ${shouldHighlight} and hasModalInformation is ${hasModalInformation}`, () => {
        it(`should return ${result}`, () => {
          expect(getClassName(shouldHighlight, hasModalInformation)).toBe(result);
        });
      });
    });
  });
});
