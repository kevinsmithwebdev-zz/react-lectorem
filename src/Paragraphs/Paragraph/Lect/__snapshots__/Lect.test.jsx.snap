// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lect the main component render empty should render without crashing 1`] = `null`;

exports[`Lect the main component render with data, should highlight should render and match snapshot 1`] = `
<span
  aria-hidden="true"
  className="Lect highlighted"
  onClick={[Function]}
>
  Hola.  
</span>
`;

exports[`Lect the main component render with data, should not highlight should render and match snapshot 1`] = `
<span
  aria-hidden="true"
  className="Lect unHighlighted"
  onClick={[Function]}
>
  Hola.  
</span>
`;
